New implementation of DisplayPlot based on XChart

akofke opened the pull request:
Some remaining questions:

- ~Zoom on both X and Y axes, and keep zoom when chart is updated~

- ~Draw discontinuous lines~

akofke committed:
WIP start on new implementation of DisplayPlot using XChart

akofke committed:
Basic working implementation of DisplayPlot with XChart

Modify LJMC3DDimer to have additional panel testing it.

akofke committed:
Add more features to XChart DisplayPlot

Associate each series with a String key to allow
retrieving series and setting properties. Add a name to
DataSetSink for storing the key.

akofke committed:
Filter points from plot which have a NaN y-value

Convert a GlassGraphic plot for testing.

ajschult commented:
Another:

- how does it handle plots with lots of points (how many points can we have before it gets bogged down).  plots typically have ~100 points per series, which ptplot handles fine, but it'd be nice to have more sometimes.  I think ptplot starts getting bogged down around 1000.  you can change the number of points held by one of the histories to increase this arbitrarily high.

Is our migration plan to leave in ptplot and migrate graphical simulations over when we're bored?

akofke committed:
Add alternative API to XChart DisplayPlot compatible with current

Convert some charts in GlassGraphic to test compatibility API.

akofke committed:
Invoke XYSeries update on EDT to prevent race condition with rapid data changes

akofke committed:
More XChart DisplayPlot features and bug fixes

Trim NaNs from head and tail of the series but not in the middle, this allows
line plot discontinuities where NaNs are present. Implement zooming with a LayerUI.

akofke committed:
Use double[] instead of list for passing filtered data to XChart

Much faster and more memory-efficient for large numbers of data points since
primitive arrays don't need to be copied again by XChart.

akofke committed:
Use custom JPanel for displaying chart

Allows shift + mouse wheel to zoom, context menu with options for reset zoom,
toggle log, display values, etc.

akofke committed:
Fix index out of bounds in DataSet

akofke committed:
Use setPreferredSize since setSize does nothing

akofke committed:
Convert some modules to use new XChart DisplayPlot

This commit is mostly just the ones that can be ported as-is.

akofke committed:
Convert SamGraphic to new DisplayPlotXChart

Use HistoryCollapsingDiscard to avoid unlimited
data accumulation

akofke committed:
Convert DevicePlotPoints and B2Fit to use XChart

Modify SimulationGraphic to have LaF initialization in a static
method to allow using it in other graphical apps

akofke committed:
Don't use cached arrays for filtering data in DisplayPlotXChart

Causes a race condition where arrays are re-updated while they're
being drawn. For now just make a new array for every update.

akofke committed:
Convert PistonCylinderGraphic to new DisplayPlot

Lag with rapid updates should be fixed.

akofke committed:
Convert StatisticsMCGraphic to use new DisplayPlot
